### ISSUE TITLE
Dynamic GPU support

### ISSUE THREAD

**Issue by thhart** [microsoft/onnxruntime#7044]
**Is your feature request related to a problem? Please describe.**
I am using the library on various systems which might have GPU and/or the necessary CUDA libraries installed. So I would like to have a possibility let onnxruntime decide which device to use.

As soon as I ship onnxruntime with GPU support (taken from Maven) on systems where GPU is not available or not installed CUDA wise onnxruntime fails to start.

**System information**
1.7.0

**Describe the solution you'd like**
Onnxruntime should have following behavior:
1. Check to load GPU libraries
2. Check to have GPU device available
3. Fallback to CPU if 1 or 2 failed

**Comment by Craigacp** [comment]
The Java GPU artifact is linked against CUDA which means it triggers a linker error when it fails to resolve some of the symbols when CUDA isn't available. I think the ORT team are considering making the CUDA execution provider packaged into a separate library, which could then be dependently loaded if CUDA is available and if it's not it would throw a different linker error which could be caught and fall back to CPU execution. Checking to see if the GPU device is available when CUDA is present is a little trickier, but the CUDA execution provider itself might make that check.

However at the moment the Java API cannot load any execution providers which are packaged in separate libraries due to a hard coded path in the ORT library loader and a lack of an appropriate entry point for Java, so we'd need to fix that first (tracked in #6553).

**Comment by thhart** [comment]
Thanks for picking up, meanwhile I will separate the libraries and make a GPU detector before executing Java. Just an idea, maybe it is wise to introduce an environment variable ONNX_PROVIDER_SELECTOR = CPU | GPU | AUTOMATIC to force the library loading and device detection, AUTOMATIC support might be postponed then until fully implemented.

**Comment by Craigacp** [comment]
We can't do that at the moment without doubling the size of the jar as it would require shipping both a GPU and non-GPU native library in the same jar. If CUDA is broken out into a separate provider then as I mentioned above a bunch of work needs to happen before the Java binding will support GPUs again.

However you can currently control this by overriding the library loading, if you have the native libraries unpacked on disk somewhere you can tell the JVM to load either the CPU or GPU native libraries (using the same jar files) by setting the appropriate `onnxruntime.native.{onnxruntime,onnxruntime4j_jni}.path` variable on startup. The library loading logic is here - https://github.com/microsoft/onnxruntime/blob/master/java/src/main/java/ai/onnxruntime/OnnxRuntime.java#L156.

**Comment by Craigacp** [comment]
Note for the fix in #8013 you'll need to detect if you have a GPU manually (one method could be shell calling out to see if `nvidia-smi` is present, or if `lsmod | grep nvidia` returns anything, but it's going to be platform dependent) and then only load the CUDA provider if you do, but at least you'll be able to use the onnxruntime_gpu jar everywhere.
